src/bedrock/hash_ring.cpp

```diff
@@ -0,0 +1,212 @@
+// metadata flag = 0 means the key is a metadata. Otherwise, it is a regular data
+unordered_set<server_thread_t, thread_hash> get_responsible_threads(
+    string respond_address,
+    string key,
+    bool metadata,
+    unordered_map<unsigned, global_hash_t>& global_hash_ring_map,
+    unordered_map<unsigned, local_hash_t>& local_hash_ring_map,
+    unordered_map<string, key_info>& placement,
+    SocketCache& pushers,
+    vector<unsigned>& tier_ids,
+    bool& succeed,
+    unsigned& seed) {
+
+  if (metadata) {
+    succeed = true;
+    return get_responsible_threads_metadata(key, global_hash_ring_map[1], local_hash_ring_map[1]);
+  } else {
+    unordered_set<server_thread_t, thread_hash> result;
+
+    if (placement.find(key) == placement.end()) {
+      issue_replication_factor_request(respond_address, key, global_hash_ring_map[1], local_hash_ring_map[1], pushers, seed);
+      succeed = false;
+    } else {
+      for (auto id_iter = tier_ids.begin(); id_iter != tier_ids.end(); id_iter++) {
+        unsigned tier_id = *id_iter;
+        auto mts = responsible_global(key, placement[key].global_replication_map_[tier_id], global_hash_ring_map[tier_id]);
+
+        for (auto it = mts.begin(); it != mts.end(); it++) {
+          string ip = it->get_ip();
+          auto tids = responsible_local(key, placement[key].local_replication_map_[tier_id], local_hash_ring_map[tier_id]);
+
+          for (auto iter = tids.begin(); iter != tids.end(); iter++) {
+            result.insert(server_thread_t(ip, *iter));
+          }
+        }
+      }
+
+      succeed = true;
+    }
+
+    return result;
+  }
+}
+
+
+// query the routing for a key and return all address
+vector<string> get_address_from_routing(
+    user_thread_t& ut,
+    string key,
+    zmq::socket_t& sending_socket,
+    zmq::socket_t& receiving_socket,
+    bool& succeed,
+    string& ip,
+    unsigned& thread_id,
+    unsigned& rid) {
+
+  communication::Key_Request key_req;
+  key_req.set_respond_address(ut.get_key_address_connect_addr());
+  key_req.add_keys(key);
+
+  string req_id = ip + ":" + to_string(thread_id) + "_" + to_string(rid);
+  key_req.set_request_id(req_id);
+  rid += 1;
+
+  // query routing for addresses on the other tier
+  auto key_response = send_request<communication::Key_Request, communication::Key_Response>(key_req, sending_socket, receiving_socket, succeed);
+  vector<string> result;
+
+  if (succeed) {
+    for (int j = 0; j < key_response.tuple(0).addresses_size(); j++) {
+      result.push_back(key_response.tuple(0).addresses(j));
+    }
+  }
+
+  return result;
+}
+
+routing_thread_t get_random_routing_thread(vector<string>& routing_address, unsigned& seed) {
+  string routing_ip = routing_address[rand_r(&seed) % routing_address.size()];
+  unsigned tid = rand_r(&seed) % PROXY_THREAD_NUM;
+  return routing_thread_t(routing_ip, tid);
+}
+
+
+void warmup(unordered_map<string, key_info>& placement) {
```

Comment:
same comment as above

Reply:
This one I agree could go somewhere else. Is it solely for benchmark use? If so I can move it there.

Reply:
Yes, this warmup is solely for benchmark use

Reply:
inlined in common.h